### PR TITLE
Update custom derivatives to new API

### PR DESCRIPTION
**Pull request by NeilGirdhar**
Addresses #10

### CODE REVIEW

**Comment by NeilGirdhar** [comment]
Let me know if you want to switch to poetry and I'll put the version requirements into pyproject.toml.  Otherwise, I'll update setup.py.

**Comment by gehring** [comment]
@pierrelux Any thoughts on using [poetry](https://python-poetry.org/) to replace setup.py?

**Comment by gehring** [comment]
I can't get this to run. I suspect the issue is that the new defvjp API expects a function to be called during the forward pass but None is given instead.

**Comment by NeilGirdhar** [comment]
@gehring Yes, I noticed that you weren't passing it before, and wasn't sure.  I'll look into it, but part of the problem is that the tests fail for me even without this change.  Do the tests work on your end?

**Comment by gehring** [comment]
Just ran (cpu) `two_phase_test.py` on master without any issues. If tests are failing for you, it might be worth it to open an issue to help me track down what might be causing it.

**Comment by NeilGirdhar** [comment]
It might be cause I've updated JAX.  I'll work on this today anyway.  It works out because I'd like to learn how FAX works anyway.

**Comment by gehring** [comment]
Are you using the most recent version of jax and jaxlib? Also, what hardware are you running this on and are you running with CUDA?

**Comment by NeilGirdhar** [comment]
@gehring I'm using the bleeding edge, syncing to master since I am waiting for my PR to be accepted and I want to keep it up to date. My CUDA is disabled since my drivers keep getting messed up by Ubuntu updates.  Hopefully the upgrade to 20.04 in two weeks will sort that out.

**Comment by NeilGirdhar** [comment]
@gehring I fixed the change (can you test it out?). Also, I added the change to modern `pyproject.toml` pending Pierre-Luc's approval.

File: poetry.lock [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+optional = false
+python-versions = ">=3.5.2"
+version = "5.8.0"
+
+[package.dependencies]
+attrs = ">=19.2.0"
+sortedcontainers = ">=2.1.0,<3.0.0"
+
+[package.extras]
+all = ["django (>=1.11)", "dpcontracts (>=0.4)", "lark-parser (>=0.6.5)", "numpy (>=1.9.0)", "pandas (>=0.19)", "pytest (>=4.3)", "python-dateutil (>=1.4)", "pytz (>=2014.1)"]
```
Comment (pierrelux) [inline poetry.lock:57]: @NeilGirdhar Why are these dependencies listed here? Is it really needed or just a dump of what's on your system?

Comment (NeilGirdhar) [reply poetry.lock:57]: There's a section called "Commit your poetry.lock file to version control" under [basic usage](https://python-poetry.org/docs/basic-usage/) that explains why it's checked in. I'm going to split the two commits in this pull request into two pull requests though, so we can take a look at it there if you have other questions.

File: fax/implicit/twophase.py [hunk 2]
```diff
@@ -2,6 +2,7 @@
 import logging
 
 import jax
+from jax import numpy as np, tree_util
```
Comment (gehring) [inline fax/implicit/twophase.py:5]: Could you split this into two lines?

Comment (NeilGirdhar) [reply fax/implicit/twophase.py:5]: Yup, good catch.  I think isort prefers that.

**Pull request merged**